persistent/Database/Persist/Sql/Raw/QQ.hs

```diff
@@ -0,0 +1,62 @@
+{-# LANGUAGE TemplateHaskell #-}
+{-# LANGUAGE OverloadedStrings #-}
+
+module Database.Persist.Sql.Raw.QQ (
+      sqlQQ
+    , executeQQ
+    ) where
+
+import Control.Arrow (first, second)
+import Data.Text (pack)
+import qualified Language.Haskell.TH as TH
+import Language.Haskell.TH.Quote
+import Language.Haskell.Meta.Parse
+
+import Database.Persist.Class (toPersistValue)
+import qualified Database.Persist.Sql.Raw as Raw
+
+data StringPart
+  = Literal String
+  | AntiQuote String
+  deriving Show
+
+parseHaskell :: String -> String -> [StringPart]
+parseHaskell a []          = [Literal (reverse a)]
+parseHaskell a ('\\':x:xs) = parseHaskell (x:a) xs
+parseHaskell a ['\\']      = parseHaskell ('\\':a) []
+parseHaskell a ('}':xs)    = AntiQuote (reverse a) : parseStr [] xs
+parseHaskell a (x:xs)      = parseHaskell (x:a) xs
+
+parseStr :: String -> String -> [StringPart]
+parseStr a []           = [Literal (reverse a)]
+parseStr a ('\\':x:xs)  = parseStr (x:a) xs
+parseStr a ['\\']       = parseStr ('\\':a) []
+parseStr a ('#':'{':xs) = Literal (reverse a) : parseHaskell [] xs
+parseStr a (x:xs)       = parseStr (x:a) xs
+
+makeExpr :: TH.Q TH.Exp -> [StringPart] -> TH.ExpQ
+makeExpr n s = TH.appE (TH.appE [| flip (.) (first pack) |] (TH.appE [| uncurry |] n)) (go s)
```

Comment:
> Thanks @cdepillabout I managed to implement your advice, but I have doubts about the readability of this

Yeah, I think you're right.  I haven't taken a hard look at the code, but it seems like you might not need to split the `(frist pack)` and `uncurry` stuff into two separate expressions?

Would something like the following work?  I've checked to make sure that this compiles, although I'm not sure if it passes the tests:

```haskell
makeExpr :: TH.ExpQ -> [StringPart] -> TH.ExpQ
makeExpr f s = TH.appE [| uncurry $(f) . first pack |] (go s)
    where
    go [] = [| (mempty, []) |]
    go (Literal a:xs)   = TH.appE [| first (a ++) |] (go xs)
    go (AntiQuote a:xs) = TH.appE [| first ("?" ++) . second (toPersistValue $(reify a) :) |] (go xs)

```

Or maybe you could even throw the `go` function inside the quasiquotes:

```haskell
makeExpr :: TH.ExpQ -> [StringPart] -> TH.ExpQ
makeExpr f s = [| uncurry $(f) $ first pack ( $(go s) ) |]
    where
    go [] = [| (mempty, []) |]
    go (Literal a:xs)   = TH.appE [| first (a ++) |] (go xs)
    go (AntiQuote a:xs) = TH.appE [| first ("?" ++) . second (toPersistValue $(reify a) :) |] (go xs)
```